src/main/io/gps.c

```diff
@@ -960,30 +1419,54 @@ void gpsUpdate(timeUs_t currentTimeUs)
     if (executeTimeUs > gpsStateDurationUs[gpsCurrentState]) {
         gpsStateDurationUs[gpsCurrentState] = executeTimeUs;
     }
+
     schedulerSetNextStateTime(gpsStateDurationUs[gpsData.state]);
 }
 
 static void gpsNewData(uint16_t c)
 {
     if (!gpsNewFrame(c)) {
+        // no new nav solution data
         return;
     }
 
     if (gpsData.state == GPS_STATE_RECEIVING_DATA) {
-        // new data received and parsed, we're in business
-        gpsData.lastLastMessage = gpsData.lastMessage;
-        gpsData.lastMessage = millis();
+        // set the time interval between when we last processed a Nav message.
+        // necessary for NMEA since we do not get a millisecond accurate time stamp with NMEA
+        const uint32_t now = millis();
+        gpsData.navMessageIntervalMs = now - gpsData.lastNavMessage;
+        gpsData.lastNavMessage = now;
+
+#ifdef USE_GPS_UBLOX
+        // Use gpsSol time stamp value from UBX nav solution packets when we get a valid nav solution to process
+        // calculate the interval, handling iTow wraparound at the end of the week
+        const uint32_t weekDurationMs = 604800000; // Number of milliseconds in a week
+        gpsData.gpsNavSolIntervalMs = (gpsSol.time - gpsData.lastNavSolTs + weekDurationMs) % weekDurationMs;
+        gpsData.lastNavSolTs = gpsSol.time;
+        // constrain the interval to 20hz or a value around where we would get a connection failure
+        gpsData.gpsNavSolIntervalMs = constrain(gpsData.gpsNavSolIntervalMs, 50, 2500);
```

Comment:
(It is beter to write only valid values into public structures)
```suggestion
        int delta = (gpsSol.time - gpsData.lastNavSolTs + weekDurationMs) % weekDurationMs;
        gpsData.lastNavSolTs = gpsSol.time;
        // constrain the interval to 20hz or a value around where we would get a connection failure
        gpsData.gpsNavSolIntervalMs = constrain(delta, 50, 2500);
```

Reply:
Mind you, this runs for every single byte we read from serial. Cannot be right.

Reply:
@ledvinap  Fair we have a guard on 1428

Reply:
Thanks, it does run every nav frame.
I'll make the suggested change.
The only parameter I need to end up with is the delta interval, for use in GPS Rescue and in the GPS Lap timer.
I'd like to get a similar delta value from NMEA, and then share it.  Some other time calculations in gps.c and the rescue and lap timer code will be simplified.